app/lib/three_scale/jobs.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+# Periodic tasks code are extracted to constants so that
+# it can be easily run by tests.
+#
+module ThreeScale
+  module Jobs
+
+    class Task
+      def initialize(object, method, *args)
+        @object = object
+        @method = method
+        @arguments = args
+      end
+
+      def name
+        [@object, @method, @arguments].join('_').tr('():.', '_')
+      end
+
+      def run
+        @object.public_send(@method, *@arguments)
+      end
+
+      def to_s
+        "#{self.class} with: #{init_args.inspect}"
+      end
+
+      def self.map(tasks)
+        tasks.map {|task_args| new(*task_args)}
+      end
+
+      def serialize
+        {
+          klass: self.class.to_s,
+          init_args: init_args
+        }
+      end
+
+      def self.deserialize(hash)
+        hash['klass'].constantize.new(*YAML.load(hash['init_args']))  # rubocop:disable Security/YAMLLoad
+      end
+
+      protected
+
+      def init_args
+        YAML.dump([@object, @method, @arguments])
+      end
+    end
+
+    class RakeTask < Task
+
+      def initialize(task_name) # rubocop:disable Lint/MissingSuper
+        @task_name = task_name
```

Comment:
How about something like
```
Rails.application.load_tasks unless ::Rake.application
task = ::Rake.application[@task_name]
if task
  super(task, :invoke)
else
  # why should we allow missing tasks in cron at all?
  super(Rails.logger, :warn, "CRON: Rails task #{task_name} not defined")
end
```

then run would be
```
def run
  super
ensure
  @object.try :reenable
end
```

Reply:
Not sure how it's gonna work with serialization though.

Reply:
I do not get you, this looks more complicated to *read* than previous implementation with separate purposes. 

A general `Task` that takes serializable objects, another `RakeTask` that takes the name of a rake task (maybe I could add arguments of the rake task but YAGNI)

The idea is to use simple basic objects, Rails.logger is really not serializable at all (because it holds an IO)

Reply:
The point is to avoid instantiating tasks that do not exist. But this complicates things as you rightfully point out.

Perhaps at least add logging of missing tasks inside `#run` instead of just returning?